src/toxcore/dht/server/ping.rs

```diff
@@ -0,0 +1,236 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Module for PingRequest
+*/
+
+use std::time::{Duration, Instant};
+use std::mem;
+
+use futures::{future, stream, Stream};
+use parking_lot::RwLock;
+
+use toxcore::dht::packed_node::*;
+use toxcore::dht::kbucket::*;
+use toxcore::dht::server::*;
+use toxcore::dht::dht_node::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Hold data for sending PingRequest
+pub struct Ping {
+    last_time_send_ping: Instant,
+    nodes_to_send_ping: RwLock<Bucket>,
+}
+
+impl Ping {
+    /// new Ping object
+    pub fn new() -> Self {
+        Ping {
+            last_time_send_ping: Instant::now(),
+            nodes_to_send_ping: RwLock::new(Bucket::new(None)),
+        }
+    }
+
+    fn is_bad_node_timed_out(node: &DhtNode, server: &Server) -> bool {
+        node.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+    }
+
+    fn is_friend(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter().any(|friend| friend.pk == node.pk)
+    }
+
+    fn is_in_close_list(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter()
+            .map(|friend| friend.close_nodes.nodes.iter().any(|peer| peer.pk == node.pk))
+            .any(|result_bool| result_bool)
+    }
+
+    fn is_in_ping_list(node: &PackedNode, ping_list: &Bucket) -> bool {
```

Comment:
It would be better if this function accepted `self`.

Reply:
I wouldn't agree - there may be cases when self is mutable borrowed so we can't pass it as an argument and can't use simple functions that don't require it.

Reply:
I disagree. There may be cases but now there is no one. So please change it to member function

Reply:
Could you explain why at least?